### ISSUE TITLE
bloc_test - let `expect` accept `matcher` instead of `Iterable`

### ISSUE THREAD

**Issue by avi-mastov** [felangel/bloc#1935]
**Is your feature request related to a problem? Please describe.**
The expected states are fixes. Sometimes I'm just interested in the last state and I don't care about the states before it, or I just need to know that certain states were emitted, but I don't care if more states were emitted between them.

**Describe the solution you'd like**
I saw in the source that the `Iterable` passed to `expect` is just passed as is to `test.expect` which accepts a matcher. I'd like to pass a matcher where I can use `containsAll` or `containsAllInOrder` or just `contains`... you get the point.

**Describe alternatives you've considered**
The alternative is to use `test.expect` directly, because `bloc_test` can't be used with these matchers because of type mismatch.

**Additional context**
[-]

**Comment by felangel** [comment]
Hi @avi-mastov 👋 
Thanks for opening an issue!

You can currently use `expect` in the `verify` callback of `blocTest` like:

```dart
blocTest<CounterBloc, int>(
    'emits [success] when MyEvent.implemented is added',
    build: () => CounterBloc(),
    act: (bloc) => bloc.add(CounterEvent.increment),
    verify: (bloc) => expect(bloc.state, equals(1)),
);
```

If you feel that is not sufficient can you please provide a sample app that illustrates the issue and some more context around the use-case you have in mind? Thanks! 🙏

**Comment by avi-mastov** [comment]
Hi @felangel thanks for your answer!

I guess I can use your suggestion. I thought that `verify` was intended for invoking Mockito's `verify` and separate it from the `expect` callback which is dedicated for state matching. I thought it would be nice to use the `expect` parameter like the `expect` function that accepts a matcher, and it can be achieved if you only change its type from `Iterable` to `dynamic`. Do you think it'd be a bad idea?

**Comment by felangel** [comment]
Hi @avi-mastov 👋 
The reason why I would prefer not to change `expect` to accept a matcher is because that would be a breaking change and in most cases I would like to expect that all states are correct. Can you describe what the use-case is? Why do you only care about a subset of the emitted states?

**Comment by avi-mastov** [comment]
The use case is like this:
I have a BLoC, b1, that subscribes to blocs b2,b3,...bN that are its dependencies (mocked in the test). Upon their changes, b1 sends itself corresponding events, and b1's state accumulates data from these events. The test needs to verify that eventually all the accumulated data from b2 and b3 is present in b1's state, but it should be agnostic to (1) the order in which the data came, (2) other data that may have possibly been added by b4,...bN (is checked by other tests).

Regarding (2), this can be accomplished by wrapping the state itself inside a matcher, which is supported by bloc_test. However, (1) is not currently possible in `expect`, because I have to provide each intermediate state and in the exact order.

You are correct that if I care only about the last state, I can check it in the verify bloc. Maybe it's not that important. I just thought it would be nice to support a matcher like the regular test package does. It doesn't have to be a breaking change. `expect` can check the type of the argument. If it's an Iterable, it can preserve the original behaviour, and otherwise pass it as a matcher. But it's really nice to have.

Have a great day!

**Comment by felangel** [comment]
This is now available in [v8.0.0-nullsafety.0](https://pub.dev/packages/bloc_test/versions/8.0.0-nullsafety.0) 🎉

**Comment by avi-mastov** [comment]
Thank you very much, @felangel !
